from snakescale import scale

output_prefix = 'data/result'

rule dwgsim:
    input: 'data/reference.fa'
    output: 
        output_prefix + '.read1.fastq.gz',
        output_prefix + '.read2.fastq.gz',
        output_prefix + '.mutations.txt',
        output_prefix + '.mutations.vcf',
    params:
        output_prefix=output_prefix,
        r1=141,      # Length of read one (flag `-1`)
        r2=141,      # Length of read two (flag `-2`)
        c=0,         # Generate reads for platform `Illumina`
        o=1,         # Output type of FASTQ `per-read-end`
        d=150,       # Outer distance between two ends for pairs
        s=50,        # Standard deviation of the distance between pairs
        C=100,       # Mean coverage across available positions
        e=0.003,     # Error rate per base in first read
        E=0.004,     # Error rate per base in second read
        r=0.001,     # Mutation rate
        F=1e-7,      # Somatic mutation rate
        R=0,         # Fraction of mutations that are indel
        X=0,         # Probability an indel is extended
        y=0.01,      # Probability of a random DNA read
        n=10,        # Maximum number of Ns allowed in a given read
        z=42,        # Random seed
    log: 'data/result.log'
    wrapper: scale('dwgsim')
